### PR TITLE
BEMHTML: Omit optional closing tags

### PR DESCRIPTION
**Pull request by vithar**

### CODE REVIEW

File: lib/bemhtml/index.js [hunk 1]
```diff
@@ -180,7 +180,16 @@ BEMHTML.prototype.renderClose = function renderClose(prefix,
     if (content || content === 0)
       out += this.renderContent(content, isBEM);
 
-    out += '</' + tag + '>';
+    var optionalClosingTags = [
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: предалагаю оформить в виде хеша по аналогии с https://github.com/bem/bem-xjst/blob/master/lib/bemxjst/utils.js#L43-L46, чтобы получить O(1).

File: test/bemhtml-test.js [hunk 2]
```diff
@@ -104,4 +104,15 @@ describe('BEMHTML engine tests', function() {
       block: 'b1'
     }, '<div class="b1">ok</div>');
   });
+
+  describe('omit optional', function() {
+    it('closing tags', function() {
```
Comment (zxqfox) [inline test/bemhtml-test.js]: describe in optional mode
it should omit optional tags ← current test
it should not omit not optional (opposing) ← missed

File: test/bemcontext-islast-test.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ describe('BEMContext this.isLast()', function() {
         block: 'b1',
         content: 'last'
       }
-    ], '<table><tr class="b1"><td></td><td></td></tr></table>' +
+    ], '<table><tr class="b1"><td><td></table>' +
```
Comment (zxqfox) [inline test/bemcontext-islast-test.js]: it's a major change. better to hide it under option for a while

File: test/bemhtml-test.js (hunk 2 above)
Comment (tadatuta) [inline test/bemhtml-test.js]: должно быть сформулировано в виде утверждения: «it should work somehow»

File: lib/bemhtml/index.js [hunk 4]
```diff
@@ -180,7 +180,16 @@ BEMHTML.prototype.renderClose = function renderClose(prefix,
...
+      // html4
+      'html', 'head', 'body', 'p', 'ul', 'ol', 'li', 'dt', 'dd',
+      'colgroup', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td', 'option',
+      // html5
+      /* dl — Neither tag is omissible */ 'rb', 'rt', 'rtc', 'rp', 'optgroup'
+    ];
+
+    if (optionalClosingTags.indexOf(tag) === -1)
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: спрятать за опцию и по умолчанию сохранить старое поведение

File: test/bemhtml-test.js [hunk 5]
```diff
@@ -104,4 +104,15 @@ describe('BEMHTML engine tests', function() {
       block: 'b1'
     }, '<div class="b1">ok</div>');
   });
+
+  describe('omit optional', function() {
```
Comment (tadatuta) [inline test/bemhtml-test.js]: добавить тестов на одиночный тег в BEMJSON (без пользовательских шаблонов) + тесты на разное значение опции, когда появится

**Comment by tadatuta** [comment]
нужно будет добавить описание опции в доку

File: lib/bemhtml/index.js [hunk 6]
```diff
@@ -180,7 +180,16 @@ BEMHTML.prototype.renderClose = function renderClose(prefix,
     if (content || content === 0)
       out += this.renderContent(content, isBEM);
 
-    out += '</' + tag + '>';
+    var optionalClosingTags = [
+      // html4
+      'html', 'head', 'body', 'p', 'ul', 'ol', 'li', 'dt', 'dd',
+      'colgroup', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td', 'option',
+      // html5
+      /* dl — Neither tag is omissible */ 'rb', 'rt', 'rtc', 'rp', 'optgroup'
```
Comment (miripiruni) [inline lib/bemhtml/index.js]: Please add comment with links to specification: 
1) https://www.w3.org/TR/html5/syntax.html#optional-tags
2) https://html.spec.whatwg.org/multipage/syntax.html#optional-tags

File: lib/bemhtml/index.js (hunk 1 above)
Comment (miripiruni) [inline lib/bemhtml/index.js]: `optionalClosingTags` → `optionalStartTags` (In order to HTML specifications)

**Comment by miripiruni** [comment]
If there is `optionalEndTags` how about `optionalStartTags`: html, head, body, colgroup, tbody?

I would do two hashes: 
1) `optionalElements` — for check them for start and end tags: `html`, `head`, `body`, `colgroup`, `tbody`.
2) `optionalEndTags` — for check if end tag can ommited: `p`, `ul`, `ol`, `li`, `dt`, `dd`, `thead`, `tfoot`, `tr`, `th`, `td`, `option`, `rb`, `rt`, `rtc`, `rp`, `optgroup`.

**Comment by miripiruni** [comment]
Who was the first to guess that you can even omit the quotes in attributes in some cases?

I suggest to rename this PR to “HTML compression options”.

**Comment by miripiruni** [comment]
We need benchmarks for this PR.

File: lib/bemhtml/index.js (hunk 1 above)
Comment (miripiruni) [inline lib/bemhtml/index.js]: @vithar if you will be [there](https://github.com/bem/bem-xjst/blob/master/lib/bemxjst/utils.js#L43-L46) change comment text to english please.

**Comment by vithar** [comment]
Optional start tags possible only in `bem()(false)` cases because in BEM we always have at least `class` attribute.

**Comment by vithar** [comment]
I'll continue with this PR after https://github.com/bem/bem-xjst/pull/206 will be merged.

**Comment by vithar** [comment]
> Who was the first to guess that you can even omit the quotes in attributes in some cases?
> I suggest to rename this PR to “HTML compression options”.

I suggest to omit quotes in another PR, it's more complicated task.

**Comment by miripiruni** [comment]
@vithar now you can rock all night long! https://github.com/bem/bem-xjst/pull/206

**Comment by miripiruni** [comment]
@vithar ping. You have time till tomorrow meeting :)

**Comment by miripiruni** [comment]
@vithar если хочешь, я могу забрать и доделать…

**Comment by vithar** [comment]
Да, я непонятно когда дойду.

**Comment by zxqfox** [comment]
@vithar не обязательно идти, можно по интернету

**Comment by vithar** [comment]
@miripiruni can you finish this PR?

**Comment by miripiruni** [comment]
@vithar OK.

**Pull request closed without merging**